File: src/tools/filesystem.ts
import fs from 'fs/promises';
import { existsSync } from 'fs';
import path from 'path';
import { createReadStream } from 'fs';
import { createInterface } from 'readline';
import fg from 'fast-glob';
import type { WorktreeJail } from '../worktree.js';
import type { FileLockManager } from '../locks.js';
import { ToolError } from '../worktree.js';

// 1MB limit guards against accidentally loading huge binaries or logs into the LLM context.
// Override by adjusting this constant if your workload requires larger file reads.
const MAX_FILE_SIZE = 1_048_576;

const IGNORED_DIRS = new Set([
  '.git', 'node_modules', 'build', 'dist', '.next',
  '__pycache__', '.cache', 'coverage', 'out', 'target',
  'vendor', '.venv', 'venv'
]);

export async function readFile(
  jail: WorktreeJail,
  locks: FileLockManager,
  relativePath: string
): Promise<string> {
  const absPath = jail.resolve(relativePath);
  return locks.withRead(absPath, async () => {
    let stat;
    try {
      stat = await fs.stat(absPath);
    } catch {
      throw new ToolError(`Tool Error: File not found: ${relativePath}`);
    }
    if (stat.size > MAX_FILE_SIZE) {
      throw new ToolError(
        `Tool Error: File "${relativePath}" exceeds the maximum size of 1MB (${stat.size} bytes).`
      );
    }
    return fs.readFile(absPath, 'utf-8');
  });
}

export async function readLines(
  jail: WorktreeJail,
  locks: FileLockManager,
  relativePath: string,
  startLine: number,
  endLine: number
): Promise<string> {
  const absPath = jail.resolve(relativePath);
  return locks.withRead(absPath, async () => {
    try {
      await fs.access(absPath);
    } catch {
      throw new ToolError(`Tool Error: File not found: ${relativePath}`);
    }
    return new Promise<string>((resolve, reject) => {
      const lines: string[] = [];
      let lineNum = 0;
      const rl = createInterface({
        input: createReadStream(absPath),
        crlfDelay: Infinity,
      });
      rl.on('line', (line) => {
        lineNum++;
        if (lineNum >= startLine && lineNum <= endLine) {
          lines.push(line);
        }
        if (lineNum > endLine) {
          rl.close();
        }
      });
      rl.on('close', () => resolve(lines.join('\n')));
      rl.on('error', reject);
    });
  });
}

export async function createFile(
  jail: WorktreeJail,
  locks: FileLockManager,
  relativePath: string,
  content: string
): Promise<string> {
  const absPath = jail.resolve(relativePath);
  return locks.withWrite(absPath, async () => {
    if (existsSync(absPath)) {
      throw new ToolError(`Tool Error: File already exists: ${relativePath}. Use search_and_replace to modify existing files.`);
    }
    await fs.mkdir(path.dirname(absPath), { recursive: true });
    await fs.writeFile(absPath, content, 'utf-8');
    return `File created successfully: ${relativePath}`;
  });
}

export async function listDirectory(
  jail: WorktreeJail,
  relativePath: string,
  recursive: boolean
): Promise<string> {
  const absPath = jail.resolve(relativePath);

  let stat;
  try {
    stat = await fs.stat(absPath);
  } catch {
    throw new ToolError(`Tool Error: Directory not found: ${relativePath}`);
  }
  if (!stat.isDirectory()) {
    throw new ToolError(`Tool Error: Path is not a directory: ${relativePath}`);
  }

  const ignoredPattern = `{${[...IGNORED_DIRS].join(',')}}`;

  if (recursive) {
    const entries = await fg(['**/*'], {
      cwd: absPath,
      dot: true,
      onlyFiles: false,
      ignore: [`**/${ignoredPattern}/**`, `**/${ignoredPattern}`],
    });
    if (entries.length === 0) return '(empty directory)';
    return entries.sort().join('\n');
  } else {
    const entries = await fs.readdir(absPath, { withFileTypes: true });
    const filtered = entries.filter(e => !IGNORED_DIRS.has(e.name));
    if (filtered.length === 0) return '(empty directory)';
    return filtered
      .sort((a, b) => a.name.localeCompare(b.name))
      .map(e => e.isDirectory() ? `${e.name}/` : e.name)
      .join('\n');
  }
}

function isBinaryBuffer(buffer: Buffer): boolean {
  return buffer.includes(0);
}

export async function grepSearch(
  jail: WorktreeJail,
  pattern: string,
  directory?: string,
  isRegex?: boolean
): Promise<string> {
  const searchRoot = directory ? jail.resolve(directory) : jail.root;

  const ignoredPattern = `{${[...IGNORED_DIRS].join(',')}}`;
  const files = await fg(['**/*'], {
    cwd: searchRoot,
    dot: true,
    onlyFiles: true,
    ignore: [`**/${ignoredPattern}/**`, `**/${ignoredPattern}`],
  });

  let regex: RegExp;
  try {
    regex = isRegex ? new RegExp(pattern, 'g') : new RegExp(pattern.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g');
  } catch {
    throw new ToolError(`Tool Error: Invalid regex pattern: ${pattern}`);
  }

  const results: string[] = [];

  for (const relFile of files.sort()) {
    const absFile = path.join(searchRoot, relFile);

    let buf: Buffer;
    try {
      buf = await fs.readFile(absFile) as Buffer;
    } catch {
      continue;
    }
    if (isBinaryBuffer(buf)) continue;

    const content = buf.toString('utf-8');
    const lines = content.split('\n');

    const matchingLineNums: number[] = [];
    for (let i = 0; i < lines.length; i++) {
      regex.lastIndex = 0;
      if (regex.test(lines[i])) {
        matchingLineNums.push(i);
      }
    }

    if (matchingLineNums.length === 0) continue;

    const displayPath = path.relative(jail.root, absFile);
    results.push(`\n${displayPath}:`);

    // group lines with context (2 before and after)
    const shown = new Set<number>();
    const groups: Array<{start: number, end: number}> = [];

    for (const lineNum of matchingLineNums) {
      const start = Math.max(0, lineNum - 2);
      const end = Math.min(lines.length - 1, lineNum + 2);
      groups.push({ start, end });
    }

    // merge overlapping groups
    const merged: Array<{start: number, end: number}> = [];
    for (const g of groups) {
      if (merged.length === 0 || g.start > merged[merged.length - 1].end + 1) {
        merged.push({ ...g });
      } else {
        merged[merged.length - 1].end = Math.max(merged[merged.length - 1].end, g.end);
      }
    }

    for (const group of merged) {
      if (shown.size > 0) results.push('  ---');
      for (let i = group.start; i <= group.end; i++) {
        const isMatch = matchingLineNums.includes(i);
        const prefix = isMatch ? '> ' : '  ';
        results.push(`  ${prefix}${i + 1}: ${lines[i]}`);
        shown.add(i);
      }
    }
  }

  if (results.length === 0) {
    return 'No matches found.';
  }
  return results.join('\n');
}

function normalizeForFuzzy(line: string): string {
  return line.trim().replace(/\s+/g, ' ');
}

function computeSimilarity(a: string, b: string): number {
  const normA = a.split('\n').map(normalizeForFuzzy).join('\n');
  const normB = b.split('\n').map(normalizeForFuzzy).join('\n');
  if (normA === normB) return 1.0;
  const maxLen = Math.max(normA.length, normB.length);
  if (maxLen === 0) return 1.0;
  let matches = 0;
  const aChars = normA.split('');
  const bChars = normB.split('');
  const minLen = Math.min(aChars.length, bChars.length);
  for (let i = 0; i < minLen; i++) {
    if (aChars[i] === bChars[i]) matches++;
  }
  return matches / maxLen;
}

export async function searchAndReplace(
  jail: WorktreeJail,
  locks: FileLockManager,
  relativePath: string,
  searchBlock: string,
  replaceBlock: string
): Promise<string> {
  const absPath = jail.resolve(relativePath);
  return locks.withWrite(absPath, async () => {
    let content: string;
    try {
      content = await fs.readFile(absPath, 'utf-8');
    } catch {
      throw new ToolError(`Tool Error: File not found: ${relativePath}`);
    }

    // Exact match
    const exactCount = content.split(searchBlock).length - 1;
    if (exactCount === 1) {
      const updated = content.replace(searchBlock, replaceBlock);
      await fs.writeFile(absPath, updated, 'utf-8');
      const idx = updated.indexOf(replaceBlock);
      const lines = updated.split('\n');
      const charsBefore = updated.substring(0, idx).split('\n').length - 1;
      const replaceLines = replaceBlock.split('\n').length;
      const start = Math.max(0, charsBefore - 2);
      const end = Math.min(lines.length, charsBefore + replaceLines + 2);
      const segment = lines.slice(start, end).join('\n');
      return `File updated successfully.\n\nUpdated segment:\n${segment}`;
    }
    if (exactCount > 1) {
      throw new ToolError(
        `Tool Error: The search block matches ${exactCount} locations in the file. Please provide more context to make the match unique.`
      );
    }

    // Fuzzy matching
    const fileLines = content.split('\n');
    const searchLines = searchBlock.split('\n');
    const windowSize = searchLines.length;

    let bestScore = 0;
    let bestStart = -1;
    let bestEnd = -1;

    for (let i = 0; i <= fileLines.length - windowSize; i++) {
      const chunk = fileLines.slice(i, i + windowSize).join('\n');
      const score = computeSimilarity(searchBlock, chunk);
      if (score > bestScore) {
        bestScore = score;
        bestStart = i;
        bestEnd = i + windowSize;
      }
    }

    if (bestScore < 0.6) {
      const bestMatch = bestStart >= 0 ? fileLines.slice(bestStart, bestEnd).join('\n') : '(no close match found)';
      throw new ToolError(
        `Tool Error: Could not find a match for the search block. Best match (similarity: ${(bestScore * 100).toFixed(1)}%):\n${bestMatch}\n\nPlease verify your search block matches the file content.`
      );
    }

    // Apply fuzzy replacement
    const beforeLines = fileLines.slice(0, bestStart);
    const afterLines = fileLines.slice(bestEnd);
    const updatedLines = [...beforeLines, ...replaceBlock.split('\n'), ...afterLines];
    const updated = updatedLines.join('\n');
    await fs.writeFile(absPath, updated, 'utf-8');

    const start = Math.max(0, bestStart - 2);
    const end = Math.min(updatedLines.length, bestStart + replaceBlock.split('\n').length + 2);
    const segment = updatedLines.slice(start, end).join('\n');
    return `File updated successfully (fuzzy match, similarity: ${(bestScore * 100).toFixed(1)}%).\n\nUpdated segment:\n${segment}`;
  });
}

File: internal/tools/filesystem.go
package tools

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"unicode/utf8"

	"github.com/iamangus/code-mcp/internal/locks"
	"github.com/iamangus/code-mcp/internal/worktree"
)

const MaxFileSize = 1 << 20 // 1 MiB

var ignoreDirs = map[string]bool{
	".git": true, "node_modules": true, "build": true, "dist": true,
	".next": true, "__pycache__": true, ".cache": true, "coverage": true,
	"out": true, "target": true, "vendor": true, ".venv": true, "venv": true,
}

// ReadFile reads the entire content of a file within the worktree.
func ReadFile(worktreeRoot, filePath string, lm *locks.Manager) (string, error) {
	abs, err := worktree.Resolve(worktreeRoot, filePath)
	if err != nil {
		return "", err
	}
	lm.RLock(abs)
	defer lm.RUnlock(abs)

	info, err := os.Stat(abs)
	if err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: file not found: %s", filePath)}
	}
	if info.Size() > MaxFileSize {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: file %s exceeds maximum size of 1 MiB", filePath)}
	}
	content, err := os.ReadFile(abs)
	if err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot read file: %v", err)}
	}
	return string(content), nil
}

// ReadLines reads a range of lines (1-indexed, inclusive) from a file.
func ReadLines(worktreeRoot, filePath string, startLine, endLine int, lm *locks.Manager) (string, error) {
	abs, err := worktree.Resolve(worktreeRoot, filePath)
	if err != nil {
		return "", err
	}
	lm.RLock(abs)
	defer lm.RUnlock(abs)

	f, err := os.Open(abs)
	if err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: file not found: %s", filePath)}
	}
	defer f.Close()

	if startLine < 1 {
		return "", &worktree.ToolError{Message: "Tool Error: start_line must be >= 1"}
	}
	if endLine < startLine {
		return "", &worktree.ToolError{Message: "Tool Error: end_line must be >= start_line"}
	}

	var lines []string
	scanner := bufio.NewScanner(f)
	lineNum := 0
	for scanner.Scan() {
		lineNum++
		if lineNum >= startLine && lineNum <= endLine {
			lines = append(lines, scanner.Text())
		}
		if lineNum > endLine {
			break
		}
	}
	if err := scanner.Err(); err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: error reading file: %v", err)}
	}
	if lineNum < startLine {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: start_line %d exceeds file length %d", startLine, lineNum)}
	}
	return strings.Join(lines, "\n"), nil
}

// CreateFile creates a new file at the given path with the provided content.
func CreateFile(worktreeRoot, filePath, content string, lm *locks.Manager) (string, error) {
	abs, err := worktree.Resolve(worktreeRoot, filePath)
	if err != nil {
		return "", err
	}
	lm.Lock(abs)
	defer lm.Unlock(abs)

	if _, err := os.Stat(abs); err == nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: file already exists: %s", filePath)}
	}

	if err := os.MkdirAll(filepath.Dir(abs), 0755); err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot create parent directories: %v", err)}
	}

	if err := os.WriteFile(abs, []byte(content), 0644); err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot write file: %v", err)}
	}
	return fmt.Sprintf("File created successfully: %s", abs), nil
}

// ListDirectory lists the contents of a directory, optionally recursively.
func ListDirectory(worktreeRoot, dirPath string, recursive bool, lm *locks.Manager) (string, error) {
	abs, err := worktree.Resolve(worktreeRoot, dirPath)
	if err != nil {
		return "", err
	}

	info, err := os.Stat(abs)
	if err != nil || !info.IsDir() {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: %q is not a directory", dirPath)}
	}

	var sb strings.Builder

	if recursive {
		walkErr := filepath.Walk(abs, func(path string, fi os.FileInfo, err error) error {
			if err != nil {
				return nil
			}
			if fi.IsDir() && ignoreDirs[fi.Name()] && path != abs {
				return filepath.SkipDir
			}
			rel, _ := filepath.Rel(abs, path)
			if rel == "." {
				return nil
			}
			depth := strings.Count(rel, string(filepath.Separator))
			indent := strings.Repeat("  ", depth)
			if fi.IsDir() {
				sb.WriteString(fmt.Sprintf("%s%s/\n", indent, fi.Name()))
			} else {
				sb.WriteString(fmt.Sprintf("%s%s\n", indent, fi.Name()))
			}
			return nil
		})
		if walkErr != nil {
			return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: error walking directory: %v", walkErr)}
		}
	} else {
		entries, err2 := os.ReadDir(abs)
		if err2 != nil {
			return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot read directory: %v", err2)}
		}
		for _, entry := range entries {
			if ignoreDirs[entry.Name()] {
				continue
			}
			if entry.IsDir() {
				sb.WriteString(fmt.Sprintf("  %s/\n", entry.Name()))
			} else {
				sb.WriteString(fmt.Sprintf("  %s\n", entry.Name()))
			}
		}
	}

	return sb.String(), nil
}

// GrepSearch searches for a pattern (regex or literal) within files in a directory.
func GrepSearch(worktreeRoot, query, directory string, lm *locks.Manager) (string, error) {
	searchPath := worktreeRoot
	if directory != "" {
		searchPath = directory
	}
	abs, err := worktree.Resolve(worktreeRoot, searchPath)
	if err != nil {
		return "", err
	}

	re, err := regexp.Compile(query)
	if err != nil {
		re = regexp.MustCompile(regexp.QuoteMeta(query))
	}

	var results strings.Builder

	walkErr := filepath.Walk(abs, func(path string, fi os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if fi.IsDir() {
			if ignoreDirs[fi.Name()] && path != abs {
				return filepath.SkipDir
			}
			return nil
		}

		// Read the file under a short-lived lock so we don't hold it for the
		// entire walk duration.
		data, readErr := func() ([]byte, error) {
			lm.RLock(path)
			defer lm.RUnlock(path)
			return os.ReadFile(path)
		}()
		if readErr != nil {
			return nil
		}
		if !utf8.Valid(data) {
			return nil
		}

		lines := strings.Split(string(data), "\n")
		for i, line := range lines {
			if re.MatchString(line) {
				rel, _ := filepath.Rel(worktreeRoot, path)
				// Include context: 2 lines before and after
				start := i - 2
				if start < 0 {
					start = 0
				}
				end := i + 2
				if end >= len(lines) {
					end = len(lines) - 1
				}
				for ctx := start; ctx <= end; ctx++ {
					results.WriteString(fmt.Sprintf("%s:%d: %s\n", rel, ctx+1, lines[ctx]))
				}
				results.WriteString("---\n")
			}
		}
		return nil
	})
	if walkErr != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: error walking directory: %v", walkErr)}
	}

	return results.String(), nil
}

// SearchAndReplace finds a block of text in a file and replaces it.
// It first tries an exact match, then falls back to fuzzy matching.
func SearchAndReplace(worktreeRoot, filePath, searchBlock, replaceBlock string, lm *locks.Manager) (string, error) {
	abs, err := worktree.Resolve(worktreeRoot, filePath)
	if err != nil {
		return "", err
	}
	lm.Lock(abs)
	defer lm.Unlock(abs)

	data, err := os.ReadFile(abs)
	if err != nil {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot read file: %v", err)}
	}
	content := string(data)

	// Exact match
	count := strings.Count(content, searchBlock)
	if count > 1 {
		return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: search_block matches %d locations in %s. Provide more surrounding context to disambiguate.", count, filePath)}
	}
	if count == 1 {
		newContent := strings.Replace(content, searchBlock, replaceBlock, 1)
		if err := os.WriteFile(abs, []byte(newContent), 0644); err != nil {
			return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot write file: %v", err)}
		}
		idx := strings.Index(newContent, replaceBlock)
		return buildContext(newContent, idx, replaceBlock, abs), nil
	}

	// Fuzzy match: normalize whitespace and compare line-by-line windows
	searchLines := strings.Split(searchBlock, "\n")
	fileLines := strings.Split(content, "\n")

	normalizeLines := func(lines []string) string {
		parts := make([]string, len(lines))
		for i, l := range lines {
			parts[i] = strings.Join(strings.Fields(strings.TrimSpace(l)), " ")
		}
		return strings.Join(parts, "\n")
	}

	normalizedSearch := normalizeLines(searchLines)
	windowSize := len(searchLines)

	bestRatio := 0.0
	bestStart := -1
	bestEnd := -1

	for i := 0; i <= len(fileLines)-windowSize; i++ {
		window := fileLines[i : i+windowSize]
		normalizedWindow := normalizeLines(window)
		ratio := similarity(normalizedSearch, normalizedWindow)
		if ratio > bestRatio {
			bestRatio = ratio
			bestStart = i
			bestEnd = i + windowSize
		}
	}

	if bestRatio >= 0.85 {
		var newLines []string
		newLines = append(newLines, fileLines[:bestStart]...)
		newLines = append(newLines, strings.Split(replaceBlock, "\n")...)
		newLines = append(newLines, fileLines[bestEnd:]...)
		newContent := strings.Join(newLines, "\n")
		if err := os.WriteFile(abs, []byte(newContent), 0644); err != nil {
			return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: cannot write file: %v", err)}
		}
		idx := strings.Index(newContent, replaceBlock)
		return buildContext(newContent, idx, replaceBlock, abs), nil
	}

	snippet := ""
	if bestStart >= 0 {
		snippet = strings.Join(fileLines[bestStart:bestEnd], "\n")
	}
	return "", &worktree.ToolError{Message: fmt.Sprintf("Tool Error: search_block not found in %s (best match similarity: %.2f). Closest match:\n%s", filePath, bestRatio, snippet)}
}

// buildContext returns up to ±5 lines of context around the replaced area.
func buildContext(content string, idx int, replaceBlock, filePath string) string {
	if idx < 0 {
		return fmt.Sprintf("Successfully replaced in %s", filePath)
	}
	lines := strings.Split(content, "\n")
	before := content[:idx]
	startLine := strings.Count(before, "\n")
	replaceLines := strings.Count(replaceBlock, "\n") + 1
	endLine := startLine + replaceLines

	ctxStart := startLine - 5
	if ctxStart < 0 {
		ctxStart = 0
	}
	ctxEnd := endLine + 5
	if ctxEnd > len(lines) {
		ctxEnd = len(lines)
	}

	return fmt.Sprintf("Successfully replaced in %s\n%s", filePath, strings.Join(lines[ctxStart:ctxEnd], "\n"))
}

// similarity computes 2*LCS_chars / (len(a)+len(b)).
func similarity(a, b string) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 1.0
	}
	lcs := lcsLength(a, b)
	return float64(2*lcs) / float64(len(a)+len(b))
}

// lcsLength computes the length of the longest common subsequence of two strings.
func lcsLength(a, b string) int {
	ra := []rune(a)
	rb := []rune(b)
	m, n := len(ra), len(rb)
	if m == 0 || n == 0 {
		return 0
	}
	prev := make([]int, n+1)
	curr := make([]int, n+1)
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if ra[i-1] == rb[j-1] {
				curr[j] = prev[j-1] + 1
			} else if curr[j-1] > prev[j] {
				curr[j] = curr[j-1]
			} else {
				curr[j] = prev[j]
			}
		}
		prev, curr = curr, prev
		for k := range curr {
			curr[k] = 0
		}
	}
	return prev[n]
}
